src/oom/oomd-util.c

```diff
@@ -280,6 +280,179 @@ int oomd_cgroup_kill(const char *path, bool recurse, bool dry_run) {
+        _cleanup_free_ char *hook_path = NULL;
+        int r;
+
+        assert(basename);
+        assert(cparams);
+        assert(ks);
+        assert(e);
+
+        log_info("Invoking oomd prekill hook %s for cgroup %s", basename, ks->ctx->path);
+
+        hook_path = path_join(VARLINK_ADDR_PATH_PREKILL, basename);
+        if (!hook_path)
+                return log_oom_debug();
+
+        r = sd_varlink_connect_address(&link, hook_path);
+        if (r < 0)
+                return log_debug_errno(r, "Prekill socket at %s not connected, ignoring.", hook_path);
+
+        (void) sd_varlink_set_userdata(link, ks);
+        r = sd_varlink_set_description(link, "oomd prekill hook");
+        if (r < 0)
+                return log_debug_errno(r, "Failed to set varlink description: %m");
+        (void) sd_varlink_set_relative_timeout(link, ks->manager->prekill_timeout);
+
+        r = sd_varlink_attach_event(link, e, SD_EVENT_PRIORITY_NORMAL);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to attach varlink to event loop: %m");
+
+        r = sd_varlink_bind_reply(link, prekill_callback);
+        if (r < 0)
+                return log_debug_errno(r, "Failed to bind reply callback: %m");
+
+        r = sd_varlink_invoke(link, "io.systemd.oom.PreKill", cparams);
+        if (r < 0)
+                return log_warning_errno(r, "Failed to call varlink method io.systemd.oom.PreKill: %m");
+
+        TAKE_PTR(link);
+        oomd_kill_state_ref(ks);
+
+        return 0;
+}
+
+static int oomd_prekill_hook(const OomdCGroupContext *ctx, Manager *m) {
+        _cleanup_closedir_ DIR *d = NULL;
+        int r;
+
+        assert(ctx);
+        assert(m);
+
+        d = opendir(VARLINK_ADDR_PATH_PREKILL);
+        if (!d) {
+                if (errno == ENOENT) {
+                        log_info("No prekill varlink socket directory %s, ignoring.", VARLINK_ADDR_PATH_PREKILL);
+                        return 0;
+                }
+                return log_debug_errno(errno, "Failed to open prekill varlink socket directory %s: %m",
+                                         VARLINK_ADDR_PATH_PREKILL);
+        }
+
+        _cleanup_(oom_kill_state_freep) struct OomdKillState *ks = new(struct OomdKillState, 1);
+        if (!ks)
+                return log_oom_debug();
+
+        *ks = (struct OomdKillState) {
+                .n_ref = 1,
+                .manager = m,
+                .ctx = ctx,
+        };
+
+        _cleanup_(sd_json_variant_unrefp) sd_json_variant *cparams = NULL;
+        r = sd_json_buildo(&cparams, SD_JSON_BUILD_PAIR_STRING("cgroup", ctx->path));
+        if (r < 0)
+                return log_oom_debug();
+
+        set_put(m->prekill_ctxs, ks);
+
+        FOREACH_DIRENT(de, d, return -errno)
+                if (IN_SET(de->d_type, DT_SOCK, DT_UNKNOWN))
+                        (void) send_prekill_message(de->d_name, cparams, ks, m->event);
+
+        oomd_kill_state_unref(TAKE_PTR(ks));
```

Comment:
I'm confused. You're sending varlink request and killing right away?
is it a debugging code or what?

Reply:
I'm not killing, I'm unrefing.
At start refcount is set to 1, and after the loop I unref it again, so if no handlers are registered I start the killing at least once